lib/engine/game/g_steam_over_holland/game.rb

```diff
@@ -403,14 +427,52 @@ def issuable_shares(entity)
         end
 
         def redeemable_shares(entity)
-          return [] unless round.steps.find { |step| step.instance_of?(GSteamOverHolland::Step::IssueShares) }.active?
+          return [] if @round.issued_shares[entity]
 
           share_price = stock_market.find_share_price(entity, :current).price
 
           bundles_for_corporation(share_pool, entity)
             .each { |bundle| bundle.share_price = share_price }
             .reject { |bundle| entity.cash < bundle.price }
         end
+
+        def emergency_issuable_bundles(entity)
+          return [] if @round.issued_shares[entity] || entity.cash >= @depot.min_depot_price
+
+          num_shares = [entity.num_player_shares, 5 - entity.num_market_shares].min
+          bundles = bundles_for_corporation(entity, entity)
+          share_price = stock_market.find_share_price(entity, :current).price
+
+          bundles
+            .each { |bundle| bundle.share_price = share_price }
+            .reject { |bundle| bundle.num_shares > num_shares }
+        end
+
+        def upgrades_to?(from, to, _special = false, selected_company: nil)
+          return %w[SOH2 SOH3 SOH4 SOH5].include?(to.name) if from.hex.tile.name == '3'
+          return %w[SOH2 SOH3 981 991].include?(to.name) if from.hex.tile.name == '4'
+          return %w[SOH4 SOH5 981 991].include?(to.name) if from.hex.tile.name == '58'
+
+          super
+        end
```

Comment:
Instead of hardcoding all these upgrades, you could override `upgrades_to_correct_city_town?` and allow upgrades where `from.towns.size == 1 && to.towns.size == 2`.

Reply:
very cool, didn't know about that method. Thanks! Done.